### PR TITLE
O/feature/deref

### PR DESCRIPTION
**Pull request by narongdejsrn**

### CODE REVIEW

**Comment by piti118** [comment]
what does it do in this case <<A<<B1>>>>

**Comment by piti118** [comment]
I think there are two kinds of deref then.
1) Template building time deref and
2) Processing time deref.

The second one is a bit dangerous since it gives you a dynamic structure. But I can see some use there.

What I was aiming for was 1) Template building time deref. So we should make another issue one for template building time deref.

**Review by piti118 — CHANGES_REQUESTED** [review]
Shouldn't be written like this a' This requires you to add this function to every new locator, parser making it a leaky abstraction.

This step should be a post processing for spec before putting it making an ExcelProcessor.
Eg.
```
def process(wb)
    derefed_spec = spec.deref_everything(wb)
    processor = ExcelProcessor.from_spec(derefed_spec)
    ...
```

**Comment by narongdejsrn** [comment]
> what does it do in this case `<<A<<B1>>>>`
> 
> The question is should it be done recursively??? Can't think of a good use case for now though.
> 
> There is also a danger of infinite loop like if A1 contains text like `<<A1>>`.
> 
> So I think deref once strategy is good for now.

Wouldn't work, only deref once (covered by test)

> Shouldn't be written like this a' This requires you to add this function to every new locator, parser making it a leaky abstraction.
> 
> This step should be a post processing for spec before putting it making an ExcelProcessor.
> Eg.
> 
> ```
> def process(wb)
>     derefed_spec = spec.deref_everything(wb)
>     processor = ExcelProcessor.from_spec(derefed_spec)
>     ...
> ```

Fixed, and now you can turn deref off for specific block by using `deref: no`

File: src/exco/extractor/cell_extraction_task.py [hunk 1]
```diff
@@ -91,15 +92,25 @@ def __str__(self):
...
         assumptions: {[dict(key=key, a=a) for key, a in self.assumptions.items()]}
-        fallback: {self.fallback}""")
+        fallback: {self.fallback}
+        deref: {self.deref}
+        """)
         return s
 
     def process(self, anchor_cell_location: CellLocation,
                 workbook: Workbook) -> CellExtractionTaskResult[T]:
+        if self.deref:
```
Comment (piti118) [inline src/exco/extractor/cell_extraction_task.py:102]: mutating state here is a bad idea. The object can't be reused predictably a' and even worse in multithreaded environment.

File: src/exco/extractor/deref/built_in/everything_deref.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+                  workbook: Workbook,
+                  sheet_name: str) -> CellExtractionTask:
+        dc = DerefCell(workbook=workbook,
+                       sheet_name=sheet_name)
+
+        cell_extraction_task.key = dc.deref_text(cell_extraction_task.key)
+        cell_extraction_task.fallback = dc.deref_text(cell_extraction_task.fallback,
+                                                      parser=cell_extraction_task.parser)
+
+        if hasattr(cell_extraction_task.locator, 'label'):
```
Comment (piti118) [inline src/exco/extractor/deref/built_in/everything_deref.py:22]: this is also a bad idea the deref shouldn't be fixed to ```label``` and such

**Review by piti118 — COMMENTED** [review]
I'll do it myself dee gwa this design isn't extensible a'.

**Pull request closed without merging**